packer/registry_builder.go

```diff
@@ -0,0 +1,106 @@
+package packer
+
+import (
+	"context"
+	"log"
+	"strings"
+
+	"github.com/hashicorp/hcp-sdk-go/clients/cloud-packer-service/preview/2021-04-30/models"
+	packersdk "github.com/hashicorp/packer-plugin-sdk/packer"
+	packerregistry "github.com/hashicorp/packer/internal/packer_registry"
+)
+
+type RegistryBuilder struct {
+	Name                      string
+	ArtifactMetadataPublisher *packerregistry.Bucket
+
+	packersdk.Builder
+}
+
+func (b *RegistryBuilder) Prepare(raws ...interface{}) ([]string, []string, error) {
+	return b.Builder.Prepare(raws...)
+}
+
+// Run is where the actual build should take place. It takes a Build and a Ui.
+func (b *RegistryBuilder) Run(ctx context.Context, ui packersdk.Ui, hook packersdk.Hook) (packersdk.Artifact, error) {
+	runCompleted := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				log.Printf("[TRACE] marking build %q as cancelled in Packer registry", b.Name)
+				if err := b.ArtifactMetadataPublisher.PublishBuildStatus(context.TODO(), b.Name, models.HashicorpCloudPackerBuildStatusCANCELLED); err != nil {
+					log.Printf("[TRACE] failed to update Packer registry status for %q: %s", b.Name, err)
+				}
+				return
+			case <-runCompleted:
+				return
+			}
+		}
+	}()
+
+	if err := b.ArtifactMetadataPublisher.PublishBuildStatus(ctx, b.Name, models.HashicorpCloudPackerBuildStatusRUNNING); err != nil {
+		log.Printf("[TRACE] failed to update Packer registry status for %q: %s", b.Name, err)
+	}
+
+	artifact, err := b.Builder.Run(ctx, ui, hook)
+	if err != nil {
+		if parErr := b.ArtifactMetadataPublisher.PublishBuildStatus(ctx, b.Name, models.HashicorpCloudPackerBuildStatusFAILED); parErr != nil {
+			log.Printf("[TRACE] failed to update Packer registry status for %q: %s", b.Name, parErr)
+		}
+	}
+
+	// close chan to mark completion
+	close(runCompleted)
+
+	var artifacts []packersdk.Artifact
```

Comment:
Why is this a slice?

Reply:
Good question. This might be copy / pasta from the first version. I will remove in the next commit.